app/javascript/shopify_app/shopify_app.js

```diff
@@ -1,39 +1,26 @@
 import { getSessionToken } from "@shopify/app-bridge-utils";
 
-document.addEventListener("DOMContentLoaded", async () => {
-  var data = document.getElementById("shopify-app-init").dataset;
-  var AppBridge = window["app-bridge"];
-  var createApp = AppBridge.default;
-  window.app = createApp({
-    apiKey: data.apiKey,
-    shopOrigin: data.shopOrigin,
-  });
+const SESSION_TOKEN_REFRESH_INTERVAL = 2000; // Request a new token every 2s
 
-  var actions = AppBridge.actions;
-  var TitleBar = actions.TitleBar;
-  TitleBar.create(app, {
-    title: data.page,
-  });
+document.addEventListener("DOMContentLoaded", async function () {
+  // Initialize the Embedded App
+  initializeAppBridge();
+  initializeAppTitleBar();
 
   // Wait for a session token before trying to load an authenticated page
   await retrieveToken(app);
 
-  // Redirect to the requested page
-  Turbolinks.visit(data.loadPath);
-
   // Keep retrieving a session token periodically
   keepRetrievingToken(app);
-});
 
-async function retrieveToken(app) {
-  window.sessionToken = await getSessionToken(app);
-}
+  // Redirect to the requested page when DOM loads
+  var isInitialRedirect = true;
+  redirectThroughTurbolinks(isInitialRedirect);
 
-function keepRetrievingToken(app) {
-  setInterval(() => {
-    retrieveToken(app);
-  }, 50000);
-}
+  document.addEventListener("turbolinks:load", function (event) {
+    redirectThroughTurbolinks();
+  });
+});
 
 document.addEventListener("turbolinks:request-start", function (event) {
```

Comment:
Not sure if intentional. This is no longer nested inside the scope of DOMContentLoaded, but should be moved back.

Reply:
I've noticed that moving these event listeners into the `DCL` event leads to a state where the initial redirect to `/home` doesn't have an `Authorization` header for some reason. Moving the `turbolinks:request-start` event listener outside of DCL fixes it. 🤔  I'll need to look into this a bit more.

Reply:
Might be a timing/order-of-operation issue between when `Turbolinks.start()` runs vs when the `turbolinks:*` events are registered.